docs/proposals/20221611-workload-identity-integration.md

```diff
@@ -0,0 +1,351 @@
+			tokenFilePath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+			// see the next code section for details on this function
+			cred, err := newWorkloadIdentityCredential(tenantID, clientID, tokenFilePath, wiCredOptions)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to setup workload identity")
+			}
+
+			client := subscriptions.NewClient()
+
+			// setCredentialsForWorkloadIdentity just setups the 
+			// PublicCloud env URLs 
+			params.AzureClients.setCredentialsForWorkloadIdentity(ctx, params.AzureCluster.Spec.SubscriptionID, params.AzureCluster.Spec.AzureEnvironment)
+			client.Authorizer = azidext.NewTokenCredentialAdapter(cred, []string{"https://management.azure.com//.default"})
+			params.AzureClients.Authorizer = client.Authorizer
+
+		}
+```
+
+**NOTE:**
+`azidext.NewTokenCredentialAdapter` is used to get a authoriser in order to add to the existing code workflow to adapt an azcore.TokenCredential type to an autorest.Authorizer type.
+
+Also a go file e.g `workload_identity.go` in the `identity` package dealing with AWI functionality.
+
+```go
+
+type workloadIdentityCredential struct {
+	assertion string
+	file      string
+	cred      *azidentity.ClientAssertionCredential
+	lastRead  time.Time
+}
+
+type workloadIdentityCredentialOptions struct {
+	azcore.ClientOptions
+}
+
+func newWorkloadIdentityCredential(tenantID, clientID, file string, options *workloadIdentityCredentialOptions) (*workloadIdentityCredential, error) {
+	w := &workloadIdentityCredential{file: file}
+	cred, err := azidentity.NewClientAssertionCredential(tenantID, clientID, w.getAssertion, &azidentity.ClientAssertionCredentialOptions{ClientOptions: options.ClientOptions})
+	if err != nil {
+		return nil, err
+	}
+	w.cred = cred
+	return w, nil
+}
+
+func (w *workloadIdentityCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return w.cred.GetToken(ctx, opts)
+}
+
+func (w *workloadIdentityCredential) getAssertion(context.Context) (string, error) {
+	if now := time.Now(); w.lastRead.Add(5 * time.Minute).Before(now) {
+		content, err := os.ReadFile(w.file)
+		if err != nil {
+			return "", err
+		}
+		w.assertion = string(content)
+		w.lastRead = now
+	}
+	return w.assertion, nil
+}
+
+```
+
+
+### <a name='OpenQuestions'></a>Open Questions
+
+#### <a name='Howtomultitenancy'></a>1. How to achieve multi-tenancy?
+
+The identity is tied to service account via the client ID annotation present in the service account. Service account has one to one mapping with the pod. To have a many-to-one relationship between multiple identities and a service account, multiple federated identity credentials can be created that references the same service account.
+
+`azure.workload.identity/client-id` annotations contains the client ID of the identity that will be used.
+
+To use a different client ID, the client ID should be updated but the limitation is that at a time only one identity can be used. 
```

Comment:
This is not really a limitation. The annotated client-id is injected as an env var `AZURE_CLIENT_ID` that'll be automagically used by the SDKs when no client id is provided as part of the managed identity client initialization. If you wish to use a different identity, then you can explicitly pass the client id of the identity to the client.

xref: https://azure.github.io/azure-workload-identity/docs/faq.html#how-to-federate-multiple-identities-with-a-kubernetes-service-account

Reply:
Awesome. I will add that and will also test in the poc code that I have.

Reply:
Could we reuse the `AzureClusterIdentity`  CRD to specify which client ID to use per workload cluster?